### PR TITLE
Added Ascent plugin docs

### PR DESCRIPTION
**Pull request by WillTrojak**

### CODE REVIEW

File: doc/src/user_guide.rst [hunk 1]
```diff
@@ -1447,58 +1447,46 @@ Time average quantities. Parameterised with
 
     ``continuous`` | ``windowed``
 
-    In continuous mode each output file contains average data from
-    ``tstart`` until the current time. In windowed mode each output
-    file only contains average data for the most recent ``dt-out`` time
-    units. The default is ``windowed``.
+    Windowed outputs averages over each ``dt- out`` period. Whereas, continuous
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1450]: Revert.

File: doc/src/installation.rst [hunk 2]
```diff
@@ -186,3 +186,19 @@ directory which is on the library search path.  Alternatively, the
 paths can be specified explicitly by exporting the environment
 variables ``PYFR_METIS_LIBRARY_PATH=/path/to/libmetis.so`` and/or
 ``PYFR_SCOTCH_LIBRARY_PATH=/path/to/libscotch.so``.
+
+Conduit/Ascent
+^^^^^^^^^^^^^^
+
+Detailed instructions on the compilation and install of
```
Comment (pv101) [inline doc/src/installation.rst:193]: install -> installation

File: doc/src/user_guide.rst [hunk 3]
```diff
@@ -1570,6 +1570,80 @@ Example::
     int-E = rho*(u*u + v*v + w*w)
     int-enst = rho*(%(vor1)s*%(vor1)s + %(vor2)s*%(vor2)s + %(vor3)s*%(vor3)s)
 
+.. _ascent-plugin:
+
+[soln-plugin-ascent]
+^^^^^^^^^^^^^^^^^^^^
+
+Uses `Alpine Ascent <https://github.com/Alpine-DAV/ascent>`_ to plot on-the-fly
+. You will need both Conduit and Ascent compiled with MPI and VTK-h support. The
```
Comment (pv101) [inline doc/src/user_guide.rst:1579]: VTM-m ?

Comment (pv101) [inline doc/src/user_guide.rst:1579]: Do we need to repeat this part that is already in the installation guide?

File: doc/src/user_guide.rst [hunk 4]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+
+1. ``nsteps`` --- produce the plots every ``nsteps`` time steps:
+
+    *int*
+
+2. ``division`` --- the level of linear subdivison to use
+
+    *int*
+
+3. ``region`` --- region to integrate, specified as either the
```
Comment (pv101) [inline doc/src/user_guide.rst:1596]: Should 'integrate' be 'visualise'?

File: doc/src/user_guide.rst [hunk 5]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+define expressions for the fields used. This can either be a scalar or a
+vector which is defined by a comma separted list of expressions.
+
+    *string* | *string*, *string* (, *string*)
+
+2. ``scene-{name}`` --- a scene to plot with ascent options passed in a
+dictionary. Each scene needs a field to be specified, and the expression for
+that field must have been set either via a field command or a pipeline.
+Additionally, one or multiple ``render-{name}`` dictionaries must be defined to
+configre the rendering of the scene. Multiple render dictionaries gives
```
Comment (pv101) [inline doc/src/user_guide.rst:1617]: configure

**Comment by FreddieWitherden** [comment]
VTK-h is its own thing but I think now comes bundled with Ascent.Regards, Freddie.On 22 Jan 2023, at 18:04, Peter Vincent ***@***.***> wrote:﻿
@pv101 commented on this pull request.



In doc/src/user_guide.rst:
> @@ -1570,6 +1570,80 @@ Example::
     int-E = rho*(u*u + v*v + w*w)
     int-enst = rho*(%(vor1)s*%(vor1)s + %(vor2)s*%(vor2)s + %(vor3)s*%(vor3)s)
 
+.. _ascent-plugin:
+
+[soln-plugin-ascent]
+^^^^^^^^^^^^^^^^^^^^
+
+Uses `Alpine Ascent <https://github.com/Alpine-DAV/ascent>`_ to plot on-the-fly
+. You will need both Conduit and Ascent compiled with MPI and VTK-h support. The

VTM-m ?

File: doc/src/installation.rst [hunk 6]
```diff
@@ -186,3 +186,19 @@ directory which is on the library search path.  Alternatively, the
...
+`Conduit <https://llnl-conduit.readthedocs.io>`_ and
+`Ascent <https://ascent.readthedocs.io>`_ can be found in the respective
+documentation. MPI and VTK-m support is currently required for the plugin. In
```
Comment (pv101) [inline doc/src/installation.rst:196]: Is it worth rephrasing this part to make it clear that VTK-m must be installed?

File: doc/src/installation.rst [hunk 7]
```diff
@@ -186,3 +186,19 @@ directory which is on the library search path.  Alternatively, the
...
+order for PyFR to find the libraries it may be necessary to export the library
+paths. For example, on linux you will need::
+
+    export PYFR_CONDUIT_LIBRARY_PATH=/path/to/libconduit.so
+    export PYFR_ASCENT_MPI_LIBRARY_PATH=/path/to/libascent_mpi.so
+
+Currently the plugin requires that Ascent and Conduit are 64-bit, this is
+default when compiling in most cases.
```
Comment (pv101) [inline doc/src/installation.rst:204]: Is there a minimum version number requirement for Ascent, and thus for Conduit and VTK-m?

**Comment by FreddieWitherden** [comment]
Is it possible to build Ascent without VTK-m support? I think it is a hard dependency so not something we need to mention.On 22 Jan 2023, at 18:11, Peter Vincent ***@***.***> wrote:﻿
@pv101 commented on this pull request.



In doc/src/installation.rst:
> @@ -186,3 +186,19 @@ directory which is on the library search path.  Alternatively, the
 paths can be specified explicitly by exporting the environment
 variables ``PYFR_METIS_LIBRARY_PATH=/path/to/libmetis.so`` and/or
 ``PYFR_SCOTCH_LIBRARY_PATH=/path/to/libscotch.so``.
+
+Conduit/Ascent
+^^^^^^^^^^^^^^
+
+Detailed instructions on the compilation and install of
+`Conduit <https://llnl-conduit.readthedocs.io>`_ and
+`Ascent <https://ascent.readthedocs.io>`_ can be found in the respective
+documentation. MPI and VTK-m support is currently required for the plugin. In

Is it worth rephrasing this part to make it clear that VTK-m must be installed?

**Comment by FreddieWitherden** [comment]
Conduit is the current pre-release which will be 0.8.6, I think. Not sure about Ascent.On 22 Jan 2023, at 18:12, Peter Vincent ***@***.***> wrote:﻿
@pv101 commented on this pull request.



In doc/src/installation.rst:
> +
+Conduit/Ascent
+^^^^^^^^^^^^^^
+
+Detailed instructions on the compilation and install of
+`Conduit <https://llnl-conduit.readthedocs.io>`_ and
+`Ascent <https://ascent.readthedocs.io>`_ can be found in the respective
+documentation. MPI and VTK-m support is currently required for the plugin. In
+order for PyFR to find the libraries it may be necessary to export the library
+paths. For example, on linux you will need::
+
+    export PYFR_CONDUIT_LIBRARY_PATH=/path/to/libconduit.so
+    export PYFR_ASCENT_MPI_LIBRARY_PATH=/path/to/libascent_mpi.so
+
+Currently the plugin requires that Ascent and Conduit are 64-bit, this is
+default when compiling in most cases.

Is there a minimum version number requirement for Ascent, and thus for Conduit and VTK-m?

File: doc/src/user_guide.rst [hunk 8]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+    ``*`` | ``shape(args, ...)``
+
+There are then three components that can be used to build plots. Scenes which
+actualy define the render, Pipelines that can be used to apply filers and build
+sequenceces of data manipulations, and Fields which be used to define field
+expresions.
+
+1. ``field-{name}`` --- this is an extension to the Ascent library where users
+define expressions for the fields used. This can either be a scalar or a
+vector which is defined by a comma separted list of expressions.
```
Comment (pv101) [inline doc/src/user_guide.rst:1609]: -> or a vector, where the latter is defined by ...

File: doc/src/user_guide.rst [hunk 9]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+
+    ``*`` | ``shape(args, ...)``
+
+There are then three components that can be used to build plots. Scenes which
+actualy define the render, Pipelines that can be used to apply filers and build
+sequenceces of data manipulations, and Fields which be used to define field
+expresions.
+
+1. ``field-{name}`` --- this is an extension to the Ascent library where users
+define expressions for the fields used. This can either be a scalar or a
```
Comment (pv101) [inline doc/src/user_guide.rst:1608]: Should we make explicit what you have to play with here? Is it only the primitive variables rho, u, v, w, p? Can you use x, y, z, and t as well?

File: doc/src/user_guide.rst [hunk 10]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+vector which is defined by a comma separted list of expressions.
+
+    *string* | *string*, *string* (, *string*)
+
+2. ``scene-{name}`` --- a scene to plot with ascent options passed in a
```
Comment (pv101) [inline doc/src/user_guide.rst:1613]: ascent -> Ascent ?
Also, can 'Ascent options' be made a hyperlink?

File: doc/src/user_guide.rst [hunk 11]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+that field must have been set either via a field command or a pipeline.
+Additionally, one or multiple ``render-{name}`` dictionaries must be defined to
+configre the rendering of the scene. Multiple render dictionaries gives
+multiple views of the same scene.
+
+    *dict*
+
+3. ``pipeline-{name}`` --- a pipeline of data manipulations that can be used
+within a scene. The value is a dictionary containing the valid configuration
+options. Pipeline objects can be stacked together to form a pipeline of filters
```
Comment (pv101) [inline doc/src/user_guide.rst:1624]: Should we provide a list of the valid options, or a hyperlink to a list if on Ascent site?

File: doc/src/user_guide.rst (hunk 3 above)
Comment (WillTrojak) [reply doc/src/user_guide.rst:1579]: I'll move to the installation guide

File: doc/src/user_guide.rst [hunk 12]
```diff
@@ -1570,6 +1570,80 @@ Example::
...
+by making a list of dictionaries. Finally, the q-criterion and vorticity
+filters require that a field called velocity is defined.
+
+   *dict* | [*dict*]
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1628]: ?

Comment (WillTrojak) [reply doc/src/user_guide.rst:1621]: a list of dicts is supported, its the syntax used to stack pipelines

**Comment by FreddieWitherden** [comment]
@pv101 Does anything else need to be done here?

**Comment by FreddieWitherden** [comment]
@pv101 Just following up here.

@WillTrojak Can you rebase?

**Comment by FreddieWitherden** [comment]
Just following up here.

**Comment by FreddieWitherden** [comment]
@pv101 Just following up here.

File: doc/src/installation.rst [hunk 13]
```diff
@@ -186,3 +186,23 @@ directory which is on the library search path.  Alternatively, the
 paths can be specified explicitly by exporting the environment
 variables ``PYFR_METIS_LIBRARY_PATH=/path/to/libmetis.so`` and/or
 ``PYFR_SCOTCH_LIBRARY_PATH=/path/to/libscotch.so``.
+
+Ascent
+^^^^^^
+
+To run the Ascent plugin, MPI, VTK-m, and Conduit are required. VTK-m is a
+supplimentary VTK library with more visualisation algorithm, and  Conduit is a
```
Comment (pv101) [inline doc/src/installation.rst:194]: "more visualisation algorithm"?

Comment (pv101) [inline doc/src/installation.rst:194]: Double space before Conduit?

**Comment by pv101** [comment]
@WillTrojak @FreddieWitherden Sorry for delayed response here. I have just provided two minor English/formatting comments. Then I think it is good to merge.

**Comment by FreddieWitherden** [comment]
In addition to the default value of `serial` for the backend I believe that `openmp` is another option.  This might be useful for improving rendering performance.

**Comment by FreddieWitherden** [comment]
@pv101 Are you okay to roll this in with your documentation updates?

**Comment by pv101** [comment]
Yes, good idea. Will do.

**Pull request closed without merging**